include/pybind11/cast.h

```diff
@@ -34,6 +34,20 @@ template <typename Type> class opaque {
 
 NAMESPACE_BEGIN(detail)
 
+/// Convert from any type pointer to another avoding warnings about void* vs (void*()) incompatibility.
```

Comment:
Since all of this is just to remove a warning without changing the functionality, I don't think it's a good idea to add lots of code that makes the underlying code harder to understand.

Instead, I would propose adding some pragma akin to 
```
#  pragma GCC diagnostic ignored "-Wmissing-field-initializers"
```
(for your specific warning message) to ``pybind11.h`` and ``numpy.h``